### DISCUSSION TITLE
OpenTofu (née OpenTF) support

### DISCUSSION THREAD

**Discussion by skyzyx** [aquasecurity/trivy#5069] [Q&A] [answered]
### Question

[Originally asked in the [`aquasecurity/tfsec` issue tracker](https://github.com/aquasecurity/tfsec/issues/2095) in a simpler form.]

https://opentf.org/announcement

We want to understand how ~~tfsec~~ _Trivy_ is affected by the OpenTofu announcement, and what the plans for support are moving forward.

From a technical perspective, I expect that it will be a no-op for the time being as OpenTofu anticipates maintaining compat/interop for the foreseeable future. (However, OpenTofu has stated that _some_ functionality may be added to OpenTofu that is a _superset_ of what's in base Terraform.[[source](https://github.com/opentffoundation/roadmap/issues/24)])

From a process perspective:

1. If there is a bug in OpenTofu that does not exist in Terraform (or vice-versa), will bug reports (as they pertain to Trivy) be treated equally on both platforms?

1. Will Trivy testing be performed with code targeting both platforms?

1. For OpenTofu's potential future "superset" functionality, will that be treated as a first-class citizen by Trivy?

**Comment by AnaisUrlichs** [comment]
Hi there, thank you for raising this; it certainly is an important discussion to have. While I am not in a position to provide direct replies to your questions, I would assume that OpenTF is going to become supported as a Trivy Plugin.
This way, users can also expand its functionality as they see fit with additional plugin features https://github.com/aquasecurity?q=trivy+plugin&type=all&language=&sort=

**Reply by skyzyx** [reply]
I've been using `tfsec` for years, but had not yet had an opportunity to dig into Trivy yet despite it being on my radar for a while. This information is helpful in understanding how Trivy works by design. Thank you!

> […] I would assume that OpenTF is going to become supported as a Trivy Plugin.

This is the keyword I'm interested in. “Support” does not mean that it works. Support refers to what happens when something goes wrong; who “owns” solving the problem.

This is what I'm seeking clarity about. :)

**Comment by simar7** [comment]
Thanks for asking, it's hard to say what OpenTF will actually look like at this point of time as it's just a manifesto. As per the manifesto, it seems to have 1:1 compatibility with Terraform so my guess is that it should work out of the box. But then again I'm speculating as we don't have the source code to try it out just yet.

**Comment by nikpivkin** [comment]
Issue to track the first release of OpenTF: https://github.com/opentffoundation/roadmap/issues/10.

**Comment by nikpivkin** [comment]
OpenTF is available https://github.com/opentffoundation/opentf

**Comment by amirasyraf** [comment]
Has there been any consideration on this at all recently? This is very important.

OpenTofu is now GA: https://opentofu.org/blog/opentofu-is-going-ga/

At the moment, there is no doubt that tfsec will probably work on both without much issue. But as time goes on, the feature gap between Terraform and OpenTofu will only get wider, and will eventually be incompatible with one another.

What is the plan going forward? Will tfsec continue supporting Terraform? Will there be a OpenTofu support? Or will it support both?

**Reply by simar7** [reply]
We don't have any plans at the moment. Is there anything in particular in OpenTofu that Trivy is unable to scan? If so, please share a reproducible code example so we can evaluate.

**Reply by amirasyraf** [reply]
Hi simar7,
I haven't used Trivy, just asking for tfsec specifically as my issue in the tfsec repo was closed and was redirected here.

Just for asking for plan going forward, for the reason I mentioned above. For example, Gruntwork which develops tools like Terragrunt will only support OpenTofu going forward: https://blog.gruntwork.io/the-future-of-terraform-must-be-open-ab0b9ba65bca?gi=a713d8f1047f. Companies and organizations have to make decisions on which path they're going to take. Some may be able to support both, but I doubt most have the resource to do so once the gap between Terraform and OpenTofu widen.

The 1:1 compatibility between Terraform and OpenTofu is really for this early stage of the fork. Going forward (since the license change to BSL), Terraform isn't going to be the upstream source for OpenTofu.

**Reply by itaysk** [reply, marked as answer]
as you probably realized, tfsec has been merged into Trivy, so let's focus the discussion on Trivy.
Trivy supports several IaC languages, including Terraform. While ideally we would like it to support everything, we are limited in capacity and are prioritizing based on user feedback. If we see that enough Trivy users will use OpenTofu, and we see that Trivy has a limitation with OpenTofu, we would be able to evaluate the effort and prioritize accordingly. In the meantime, we welcome the OpenTofu community to make any suggestions or to engage in this task.

**Reply by AnaisUrlichs** [reply]
**Just marking this as answer since the thread will likely continue here, and I want new people following the discussion to see the highlighted comment
